ocaml/typing/solver_intf.ml

```diff
@@ -0,0 +1,298 @@
+     obj]". That helps reduce the information each [morph] needs to store.
+
+     As a result, in the interaction between the solver and the lattices,
+     [morph] always comes with its target object. *)
+
+  (** Give the source object of a morphism  *)
+  val src : 'b obj -> ('a, 'b, 'd) morph -> 'a obj
+
+  (** Give the identity morphism on an object *)
+  val id : ('a, 'a, 'd) morph
+
+  (** Compose two morphisms *)
+  val compose :
+    'c obj -> ('b, 'c, 'd) morph -> ('a, 'b, 'd) morph -> ('a, 'c, 'd) morph
+
+  (* The following returns weaker than what we want, which is "\exists r.
+     allowed * r". But ocaml doesn't like existentials, and this weaker version
+     is good enough for us *)
+
+  (** Give left adjoint of a morphism  *)
+  val left_adjoint :
+    'b obj -> ('a, 'b, 'l * allowed) morph -> ('b, 'a, left_only) morph
+
+  (** Give the right adjoint of a morphism *)
+  val right_adjoint :
+    'b obj -> ('a, 'b, allowed * 'r) morph -> ('b, 'a, right_only) morph
+
+  (** Forget whether a function can be on the right. *)
+  val disallow_right :
+    ('a, 'b, 'l * 'r) morph -> ('a, 'b, 'l * disallowed) morph
+
+  (** Forget whether a morphism can be on the left. *)
+  val disallow_left : ('a, 'b, 'l * 'r) morph -> ('a, 'b, disallowed * 'r) morph
+
+  (** Generalize a morphism that can be on the right   *)
+  val allow_right : ('a, 'b, 'l * allowed) morph -> ('a, 'b, 'l * 'r) morph
+
+  (** Generalize a morphism that can be on the left  *)
+  val allow_left : ('a, 'b, allowed * 'r) morph -> ('a, 'b, 'l * 'r) morph
+
+  (** Apply morphism on constant *)
+  val apply : 'b obj -> ('a, 'b, 'd) morph -> 'a -> 'b
+
+  val print_obj : Format.formatter -> 'a obj -> unit
+
+  (** Print morphism *)
+  val print_morph : 'b obj -> Format.formatter -> ('a, 'b, 'd) morph -> unit
+end
+
+module type S = sig
+  type 'a error =
+    { left : 'a;
+      right : 'a
+    }
+
+  module Solver_mono (C : Lattices_mono) : sig
+    type changes
+
+    val undo_changes : changes -> unit
+
+    val append_changes : (changes ref -> unit) ref
+
+    type ('a, 'd) mode
+
+    (** The minimum in the given object *)
+    val min : 'a C.obj -> ('a, 'l * 'r) mode
+
+    (** The maximum in the given object *)
+    val max : 'a C.obj -> ('a, 'l * 'r) mode
+
+    (** Returns a constant mode *)
+    val of_const : 'a -> ('a, 'l * 'r) mode
+
+    (** Applying morphism to mode *)
+    val apply :
+      'b C.obj ->
+      ('a, 'b, 'd0 * 'd1) C.morph ->
+      ('a, 'd0 * 'd1) mode ->
+      ('b, 'd0 * 'd1) mode
+
+    (** Force the first mode to be less than the second mode *)
+    val submode :
+      'a C.obj ->
+      ('a, allowed * 'r) mode ->
+      ('a, 'l * allowed) mode ->
+      (unit, 'a error) result
+
+    (** Create a new mode *)
+    val newvar : 'a C.obj -> ('a, 'l * 'r) mode
+
+    (** Fix a potentially floating mode to its lower bound, which is returned *)
+    val constrain_lower : 'a C.obj -> ('a, allowed * 'r) mode -> 'a
```

Comment:
I think the call site of this function will have a meaning for "default". For example, `optimize_allocation` try to make mode as `local` as possible. However, the mode solver itself doesn't need to talk about "default value" and such. How about `push_to_min`?

Reply:
GHC has a word for this: it's called zapping. That is, zapping is what you do to an insufficiently-constrained variable when you need it not to be a variable anymore but to be some constant. We have to do this within the layouts system, as well, where it's called defaulting. But I see how using default here could be confusing. I don't really like `push`. Here are some ideas:

* `constrain_to_lower_bound`: That is, constrain is OK. But `constrain_lower` makes me think that we're adding a new lower bound somehow, but that's very much not what we're doing.
* `zap_to_lower_bound`: Adopt the "zap" terminology.
* `floor`: this is essentially a round-down operation, so `floor` seems somewhat sensible (`ceil` would be the opposite operation)

I would avoid `min` here, because we're not interacting with the minimum.

Reply:
`floor` sounds good, but doesn't emphasize enough that this operation has side effects (not just returning the lower bound). How about "zap_floor"?

Reply:
Yeah, I was worried about its destructive nature being pretty hidden. I'd take `zap_to_floor` over `zap_floor`, but not strongly.

Reply:
Renamed to `zap_to_floor` and resolved.